triedb/pathdb/states.go

```diff
@@ -43,9 +49,528 @@ func (c *counter) report(count metrics.Meter, size metrics.Meter) {
+	// Load the destruct journal whose availability is always expected
+	destructs, err := s.journal.pop()
+	if err != nil {
+		panic(fmt.Sprintf("failed to revert state, %v", err))
+	}
+	// Revert the modifications to the destruct set by journal
+	var delta int
+	for _, entry := range destructs {
+		if entry.Exist {
+			continue
+		}
+		delete(s.destructSet, entry.Hash)
+		delta -= common.HashLength
+	}
+	// Overwrite the account data with original value blindly
+	for addrHash, blob := range accountOrigin {
+		if len(blob) == 0 {
+			if data, ok := s.accountData[addrHash]; ok {
+				delta -= common.HashLength + len(data)
+			} else {
+				panic(fmt.Sprintf("non-existent account for deleting, %x", addrHash))
+			}
+			delete(s.accountData, addrHash)
+		} else {
+			if data, ok := s.accountData[addrHash]; ok {
+				delta += len(blob) - len(data)
+			} else {
+				delta += len(blob) + common.HashLength
+			}
+			s.accountData[addrHash] = blob
+		}
+	}
+	// Overwrite the storage data with original value blindly
+	for addrHash, storage := range storageOrigin {
+		// It might be possible that the storage set is not existent because
+		// the whole storage is deleted.
+		slots := s.storageData[addrHash]
+		if len(slots) == 0 {
+			slots = make(map[common.Hash][]byte)
+		}
+		for storageHash, blob := range storage {
+			if len(blob) == 0 {
+				if data, ok := slots[storageHash]; ok {
+					delta -= 2*common.HashLength + len(data)
+				} else {
+					panic(fmt.Sprintf("non-existent storage slot for deleting, %x %x", addrHash, storageHash))
+				}
+				delete(slots, storageHash)
+			} else {
+				if data, ok := slots[storageHash]; ok {
+					delta += len(blob) - len(data)
+				} else {
+					delta += 2*common.HashLength + len(blob)
+				}
+				slots[storageHash] = blob
+			}
+		}
+		if len(slots) == 0 {
+			delete(s.storageData, addrHash)
+		} else {
+			s.storageData[addrHash] = slots
+		}
+	}
+	s.clearCache()
+	s.updateSize(delta)
+}
+
+// updateSize updates the total cache size by the given delta.
+func (s *stateSet) updateSize(delta int) {
+	size := int64(s.size) + int64(delta)
+	if size >= 0 {
+		s.size = uint64(size)
+		return
+	}
+	log.Error("Stateset size underflow", "prev", common.StorageSize(s.size), "delta", common.StorageSize(delta))
+	s.size = 0
+}
+
+// encode serializes the content of state set into the provided writer.
+func (s *stateSet) encode(w io.Writer) error {
+	// Encode destructs
+	destructs := make([]common.Hash, 0, len(s.destructSet))
+	for hash := range s.destructSet {
+		destructs = append(destructs, hash)
+	}
+	if err := rlp.Encode(w, destructs); err != nil {
+		return err
+	}
+	// Encode accounts
+	type Account struct {
+		Hash common.Hash
+		Blob []byte
+	}
+	accounts := make([]Account, 0, len(s.accountData))
+	for hash, blob := range s.accountData {
+		accounts = append(accounts, Account{Hash: hash, Blob: blob})
+	}
```

Comment:
this is non-deterministic. Does that matter at all? 
Or, maybe better question: can there be any upside to making the output encoding deterministic?